lang/test/org/partiql/lang/syntax/PartiQLParserSpyTest.kt

```diff
@@ -0,0 +1,183 @@
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at:
+ *
+ *      http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package org.partiql.lang.syntax
+
+import com.amazon.ion.IonSystem
+import com.amazon.ion.system.IonSystemBuilder
+import io.mockk.every
+import io.mockk.spyk
+import io.mockk.verify
+import org.antlr.v4.runtime.misc.ParseCancellationException
+import org.junit.Assert.assertEquals
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+import org.partiql.lang.domains.PartiqlAst
+
+/**
+ * This test class is used for spying the [PartiQLParser].
+ */
+internal class PartiQLParserSpyTest {
+
+    val ion: IonSystem = IonSystemBuilder.standard().build()
+    val parser = spyk<PartiQLParser>()
+
+    /**
+     * Shows that the PartiQLParser can parse different queries sequentially.
+     */
+    @org.junit.Test(expected = org.junit.Test.None::class)
+    fun twoQueries() {
+        val query00 = "SELECT a1 FROM t1"
+        val query01 = "SELECT a2 FROM t2"
+        parser.parseAstStatement(query00)
+        parser.parseAstStatement(query01)
+    }
+
+    /**
+     * Shows that the PartiQLParser can parse same queries sequentially.
+     */
+    @Test
+    fun twoSameQueries() {
+        val query00 = "SELECT a1 FROM t1"
+        val ast00 = parser.parseAstStatement(query00)
+        val ast01 = parser.parseAstStatement(query00)
+        assertEquals(ast00, ast01)
+    }
+
+    /**
+     * This test shows that SLL and LL parsing can happen sequentially
+     */
+    @Test
+    fun manuallyShowDoubleParsingSameQueriesWorks() {
+        val query00 = "SELECT a1 FROM t1"
+        val ast00 = PartiQLVisitor(ion).visit(parser.parseUsingSLL(query00))
+        val ast01 = PartiQLVisitor(ion).visit(parser.parseUsingLL(query00))
+        assertEquals(ast00, ast01)
+    }
+
+    /**
+     * This test shows that SLL and LL parsing can happen sequentially using different queries
+     */
+    @org.junit.Test(expected = org.junit.Test.None::class)
+    fun manuallyShowDoubleParsingWorks() {
+        val query00 = "SELECT a1 FROM t1"
+        val query01 = "SELECT a2 FROM t2"
+        PartiQLVisitor(ion).visit(parser.parseUsingSLL(query00))
+        PartiQLVisitor(ion).visit(parser.parseUsingLL(query01))
+    }
+
+    /**
+     * Shows that for simple queries, we'll only parse using the SLL parser.
+     */
+    @Test
+    fun sllParsesCorrectly() {
+        // Arrange
+        val query00 = "SELECT a FROM t"
+
+        // Act
+        val ast = parser.parseAstStatement(query00)
+
+        // Assert
+        verify(exactly = 1) { parser.parseUsingSLL(query00) }
```

Comment:
Rather than spying on the workings of the code in a test, it may be beneficial to instrument the parser directly emit an event when a fall-back is necessary. This could be useful not only for determining which of a raft of test cases correspond to 'tricky' queries, but also might allow those using PartiQL in applications/services to emit events/metrics themselves (perhaps as an aide to categorize latency classes, etc.)

Reply:
ACK. I'll create an issue for this.